keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,837 @@
+	// ExecConfig.ProvideClusterInfo).
+	// +optional
+	Cluster *Cluster `json:"cluster,omitempty"`
+
+	// Interactive declares whether stdin has been passed to this exec plugin.
+	Interactive bool `json:"interactive"`
+}
+
+// ExecCredentialStatus holds credentials for the transport to use.
+//
+// Token and ClientKeyData are sensitive fields. This data should only be
+// transmitted in-memory between client and exec plugin process. The Exec plugin
+// itself should at least be protected via file permissions.
+type ExecCredentialStatus struct {
+	// ExpirationTimestamp indicates a time when the provided credentials expire.
+	// +optional
+	ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+	// Token is a bearer token used by the client for request authentication.
+	Token string `json:"token,omitempty" datapolicy:"token"`
+	// PEM-encoded client TLS certificates (including intermediates, if any).
+	ClientCertificateData string `json:"clientCertificateData,omitempty"`
+	// PEM-encoded private key for the above certificate.
+	ClientKeyData string `json:"clientKeyData,omitempty" datapolicy:"security-key"`
+
+	// New:
+
+	// ProxyConfig is the config returned by the exec-proxy
+	// +optional
+	ProxyConfig *ProxyConfig `json:"proxyConfig,omitempty"`
+}
+
+
+// ProxyConfig configures the client to proxy requests via the returned address.
+type ProxyConfig struct {
+	// URL is the location of the exec proxy listener.  Supports unix domain
+	// socket locations in the format unix:///path/to/socket.
+	URL string `json: "url"`
+	// CertificateAuthorityData is required if an exec-proxy is used.
+	CertificateAuthorityData []byte `json:"certificate-authority-data"`
+}
+```
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing
+guidelines][testing-guidelines] when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+Both integration and e2e tests will be required in addition to unit tests.
+
+#### Integration
+
+Integration tests will cover:
+
+* Authentication between client and proxy.
+* Connection from client to apiserver through proxy.
+* Proxy use with kubectl (possible test location: [command-line integration
+  test
+  suite](https://github.com/kubernetes/kubernetes/tree/ec560b9737537be8c688776461bc700e8ddedb9d/test/cmd)).
+* UDS connections from the client to the proxy will be tested.
+
+#### E2E
```

Comment:
Given that this is client <-> API, an integration test should be sufficient. I would not expect that we would need to cover this with an e2e unless/until we want to add this to conformance.

Reply:
IIRC @nckturner wanted to use this feature in EKS for kubelet authn to the API server.  That may be an interesting test scenario to exercise in CI since it is a distinctly different type of client than say `kubectl`.

Reply:
> IIRC @nckturner wanted to use this feature in EKS for kubelet authn to the API server. That may be an interesting test scenario to exercise in CI since it is a distinctly different type of client than say kubectl.

That use case is correct.  We'd definitely want a scenario like this one in an E2E test.